HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReduceResource.h

```diff
@@ -0,0 +1,72 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_blas alpakaReduceResource_h
+#define HeterogeneousCore_AlpakaInterface_interface_blas_alpakaReduceResource_h
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/VecArray.h"
+
+namespace cms::alpakatools {
+  namespace reduce {
+
+
+    template<typename T, typename TAcc, typename TDevAcc, typename TQueue>
+    class ReducerResource
+    {
+      public:
+        using reduce_t = T;
+        //
+        using device_atomic_t = reduce_t;
+        using system_atomic_t = reduce_t;
+        using count_t         = std::int32_t;
+        //
+        using TBufHost      = alpaka::Buf<alpaka::DevCpu, system_atomic_t, Dim1D, Idx>;
+        using TBufAcc       = alpaka::Buf<TAcc, system_atomic_t, Dim1D, Idx>;
+
+        using TBuf1DAcc     = alpaka::Buf<TAcc, device_atomic_t, Dim1D, Idx>;
+
+        using TCountBufAcc  = alpaka::Buf<TAcc, count_t, Dim1D, Idx>;
+
+        //static ReducerResource reduction_buffers; 
+        static ReducerResource<T, TAcc, TDevAcc, TQueue>& get_reduction_resources(const TDevAcc &devAcc, TQueue queue, Idx nSrc, Idx n_blocks) {
+          static ReducerResource<T, TAcc, TDevAcc, TQueue> instance(devAcc, queue, nSrc, n_blocks);
+          return instance;
+        }
+        //     
+        auto get_host_reduce_ptr()   { return result_h.data(); }
+        auto get_device_reduce_ptr() { return result_d.data(); }
+        auto get_partial_ptr()       { return partial.data();  }
+
+        TBufHost&  get_host_reduce()   { return result_h; }
+        TBufAcc&   get_device_reduce() { return result_d; }
+        TBuf1DAcc& get_partial()       { return partial;  }
+
+        auto get_count_ptr()        { return count.data(); }
+        // 
+        auto fetch_reduce_data(TQueue queue) { alpaka::memcpy(queue, result_h, result_d); }
+
+      private: 
+        TBufHost      result_h;
+        TBufAcc       result_d;
+        TBuf1DAcc     partial;
```

Comment:
No, if I understood your point.  Let me explain this part:  purpose of ReducerResource class is to manage a single static instance of it (via get_reduction_resources) for a given set of parameters, see 
https://github.com/alexstrel/cmssw/blob/Alpaka-generic-reduction/HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h#L264
What is used in particular algorithms (or, better, compute kernels) is its view via TransformReduceArgs instance: 
https://github.com/alexstrel/cmssw/blob/Alpaka-generic-reduction/HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h#L267

Reply:
Thanks for the explanation. Unfortunately static or global storage is highly undesirable because of it effectively disallows concurrent use of these functions when the `ReducerResource<...>` type is the same.